File: src/python/command_handler.py
import logging
import asyncio
from admin_manager import is_admin, add_admin, remove_admin
from player_manager import find_player_by_username
from world_manager import set_spawn_pos
from ban_manager import add_ban, add_ip_ban, unban_user
from packets import create_chat_message_packet, create_set_spawn_position_packet, create_set_player_position_packet
from network import safe_send

async def send_private_message(player, message):
    from network import safe_send
    if player and player.ws:
        packet = create_chat_message_packet(f"&e{message}")
        await safe_send(player.ws, packet)

async def handle_command(sender_player, command_str, broadcast_func):
    if not is_admin(sender_player.username):
        await send_private_message(sender_player, "You do not have permission to use commands.")
        return

    parts = command_str.strip().split()
    command = parts[0].lower()
    args = parts[1:]

    logging.info(f"Admin '{sender_player.username}' issued command: {command_str}")

    # --- /kick <username> [reason] ---
    if command == "/kick" and len(args) >= 1:
        target_username = args[0]
        reason = " ".join(args[1:]) if len(args) > 1 else "Kicked by an admin."
        target_player = find_player_by_username(target_username)

        if target_player:
            await target_player.ws.close(1000, reason)
            await broadcast_func(create_chat_message_packet(f"&e{target_player.username} was kicked. ({reason})"))
        else:
            await send_private_message(sender_player, f"Player '{target_username}' not found.")
    
    # --- /ban <username> ---
    elif command == "/ban" and len(args) == 1:
        target_username = args[0]
        if add_ban(target_username):
            await send_private_message(sender_player, f"Player '{target_username}' has been banned.")
            target_player = find_player_by_username(target_username)
            if target_player:
                await target_player.ws.close(1000, "You have been banned.")
            await broadcast_func(create_chat_message_packet(f"&e{target_username} was banned."))
        else:
            await send_private_message(sender_player, f"Player '{target_username}' is already banned.")
            
    # --- /banip <username> ---
    elif command == "/banip" and len(args) == 1:
        target_username = args[0]
        target_player = find_player_by_username(target_username)
        if target_player:
            ip_address = target_player.ws.remote_address[0]
            if add_ip_ban(ip_address):
                await send_private_message(sender_player, f"IP {ip_address} for player '{target_username}' has been banned.")
                await target_player.ws.close(1000, "You have been IP banned.")
                await broadcast_func(create_chat_message_packet(f"&e{target_username} was IP banned."))
            else:
                await send_private_message(sender_player, f"IP {ip_address} is already banned.")
        else:
            await send_private_message(sender_player, f"Player '{target_username}' not found to get IP from.")

    # --- /broadcast <message> or /say <message> ---
    elif (command == "/broadcast" or command == "/say") and len(args) >= 1:
        message_to_broadcast = " ".join(args)
        await broadcast_func(create_chat_message_packet(f"&e{message_to_broadcast}"))
        
    elif command == "/setspawn":
        x, y, z = int(sender_player.x), int(sender_player.y), int(sender_player.z)
        rot = int(sender_player.yaw)
        
        if set_spawn_pos(x, y, z, rot):
            spawn_packet = create_set_spawn_position_packet(x, y, z, rot)
            await broadcast_func(spawn_packet)
            
            await broadcast_func(create_chat_message_packet(f"&eSpawn point updated to ({x}, {y}, {z})."))
        else:
            await send_private_message(sender_player, "Failed to set spawn point.")
            
    elif command == "/unban" and len(args) == 1:
        target_username = args[0]
        if unban_user(target_username):
            await send_private_message(sender_player, f"Player '{target_username}' unbanned.")
            await broadcast_func(create_chat_message_packet(f"&e{target_username} was unbanned."))
        else:
            await send_private_message(sender_player, f"Player '{target_username}' is not banned.")

    # --- /op <username> ---
    elif command == "/op" and len(args) == 1:
        target_username = args[0]
        if add_admin(target_username):
            await send_private_message(sender_player, f"Player '{target_username}' is now an operator.")
        
            target_player = find_player_by_username(target_username)
            if target_player:
                await send_private_message(target_player, "You are now an operator!")
        else:
            await send_private_message(sender_player, f"Player '{target_username}' is already an operator.")

    # --- /deop <username> ---
    elif command == "/deop" and len(args) == 1:
        target_username = args[0]
        if target_username.lower() == sender_player.username.lower():
             await send_private_message(sender_player, "You cannot de-op yourself.")
        elif remove_admin(target_username):
            await send_private_message(sender_player, f"Player '{target_username}' is no longer an operator.")
            target_player = find_player_by_username(target_username)
            if target_player:
                await send_private_message(target_player, "You are no longer an operator.")
        else:
            await send_private_message(sender_player, f"Player '{target_username}' is not an operator.")
    # --- /tp <who> <where> or /tp <x> <y> <z>     
    elif command == "/tp" or command == "/teleport":
        if len(args) == 2:
            tp_name = args[0]
            target_name = args[1]
            target_player = find_player_by_username(target_name)
            tp_player = find_player_by_username(tp_name)
            
            if target_player:
                sender_player.set_position(target_player.x, target_player.y, target_player.z, target_player.yaw, target_player.pitch)
                
                
                teleport_packet = create_set_player_position_packet(
                    tp_player.id, 
                    target_player.x, target_player.y + 1.6, target_player.z,
                    tp_player.yaw, tp_player.pitch, 1
                )
                
                await safe_send(sender_player.ws, teleport_packet)
                
                await broadcast_func(teleport_packet, exclude_ws=sender_player.ws)
                
                await send_private_message(sender_player, f"{tp_player.username} teleported to {target_player.username}.")
            else:
                await send_private_message(sender_player, f"Player '{target_name}' not found.")
                
        elif len(args) == 3:
            try:
                x = float(args[0])
                y = float(args[1])
                z = float(args[2])
                
                sender_player.set_position(x, y, z, sender_player.yaw, sender_player.pitch)
                
                teleport_packet = create_set_player_position_packet(
                    sender_player.id, x, y, z, sender_player.yaw, sender_player.pitch, 1
                )
                
                await safe_send(sender_player.ws, teleport_packet)
                await broadcast_func(teleport_packet, exclude_ws=sender_player.ws)
                
                await send_private_message(sender_player, f"Teleported to {x}, {y}, {z}.")
            except ValueError:
                await send_private_message(sender_player, "Invalid coordinates. Usage: /tp <x> <y> <z>")
        else:
            await send_private_message(sender_player, "Usage: /tp <player> OR /tp <x> <y> <z>")
        
    else:
        await send_private_message(sender_player, "Unknown command or invalid arguments.")

File: src/cpp/CommandHandler.cpp
#include "CommandHandler.hpp"
#include "Server.hpp"
CommandHandler::CommandHandler(Server* server) : server(server) {}
CommandHandler::~CommandHandler() {}

bool CommandHandler::executeCommand(std::string& text, int clientId, bool console) {
    std::vector<std::string> parts = this->parseText(text);

    if ((parts[0] == "/say" || parts[0] == "/broadcast") && parts.size() >= 2 && (server->adminSystem->isAdmin(clientId) || console)) {
        returnText = "";
        for (int i=1; i<parts.size(); ++i) {
            returnText += parts[i] + " ";
        }
        this->privateText = false;
        Packet chat(Protocol::Opcode::SERVER_CHAT_MESSAGE);
        chat.writeString("&e" + returnText);
        server->sendPacketToAll(chat);
        returnText = "";
        return true;
    }

    if (parts[0] == "/setspawn" && parts.size() == 1 && (server->adminSystem->isAdmin(clientId) || console)) {
        if (!console) {
            // Logger::logf(PREFIX_DEBUG, "%i called /setspawn\n");
            auto [x, y, z] = server->getClientPos(clientId);
            auto [yaw, pitch] = server->getClientRot(clientId);
            // Logger::logf(PREFIX_DEBUG, "Before level setspawn\n");
            server->level->setSpawnPos((int)x, (int)y, (int)z, (int)yaw);
            server->updateSpawnPosition();
            // Logger::logf(PREFIX_DEBUG, "After level setspawn\n");
            returnText = format("New spawn position set to %i, %i, %i", (int)x, (int)y, (int)z);
            // Logger::logf(PREFIX_DEBUG, "ReturnText: %s\n", returnText);
            this->privateText = true;
            // Logger::logf(PREFIX_DEBUG, "End setspawn\n");
        } else {
            Logger::logf(PREFIX_CC, "Can't set spawn from console!\n");
        }
        return true;
    }

    if (parts[0] == "/op" && parts.size() == 2) {
        this->privateText = true;
        if (server->adminSystem->isAdmin(clientId) || console) {
            if (server->adminSystem->op(parts[1])) {
                returnText = format("%s now operator", parts[1]);
                server->sendMsgToPlayer(server->findClientByUsername(parts[1]), "You now operator");
            } else {
                returnText = format("&c%s already operator", parts[1]);
            }
        } else {
            returnText = "&cYou cannot use this command.";
        }
        return true;
    }

    if (parts[0] == "/deop" && parts.size() == 2) {
        this->privateText = true;
        if (server->adminSystem->isAdmin(clientId) || console) {
            if (server->adminSystem->deop(parts[1])) {
                returnText = format("%s deopped", parts[1]);
                server->sendMsgToPlayer(server->findClientByUsername(parts[1]), "You are no longer an operator");
            } else {
                returnText = format("&c%s is not an operator", parts[1]);
            }
        } else {
            returnText = "&cYou cannot use this command.";
        }
        return true;
    }

    if (parts[0] == "/kick" && (parts.size() == 2 || parts.size() == 3 )) {
        this->privateText = true;
        if (server->adminSystem->isAdmin(clientId) || console) {
            if (server->adminSystem->kick(parts[1], parts.size() == 3 ? parts[2] : "")) {
                returnText = "";
                // returnText = format("%s kicked", parts[1]);
                server->sendMsgToPlayer(clientId, format("&e%s kicked", parts[1]));
                if (parts.size() == 3) server->sendMsgToPlayer(clientId, format("&eReason: %s", parts[2]));
            } else {
                returnText = format("&c%s cannot been kicked", parts[1]);
            }
        } else {
            returnText = "&cYou cannot use this command.";
        }
        return true;
    }

    if (parts[0] == "/ban" && parts.size() == 2) {
        this->privateText = true;
        if (server->adminSystem->isAdmin(clientId) || console) {
            if (server->adminSystem->ban(parts[1])) {
                returnText = format("%s banned", parts[1]);
            } else {
                returnText = format("&c%s already banned or bad username", parts[1]);
            }
        } else {
            returnText = "&cYou cannot use this command.";
        }
        return true;
    }

    if (parts[0] == "/banip" && parts.size() == 2) {
        this->privateText = true;
        if (server->adminSystem->isAdmin(clientId) || console) {
            if (server->adminSystem->banip(parts[1])) {
                returnText = format("%s banned by ip", parts[1]);
            } else {
                returnText = format("&c%s already banned by ip or bad username", parts[1]);
            }
        } else {
            returnText = "&cYou cannot use this command.";
        }
        return true;
    }

    if (parts[0] == "/tp" || parts[0] == "/teleport" && parts.size() == 2) {
        if (!console) {
            this->privateText = true;
            int otherId = server->findClientByUsername(parts[1]);
            auto [px, py, pz] = server->getClientPos(clientId);
            auto [ox, oy, oz] = server->getClientPos(otherId);
            auto [pyaw, ppitch] = server->getClientRot(clientId);
            auto [oyaw, opitch] = server->getClientRot(otherId);

            returnText = format("Teleporting to %s", parts[1]);

            Packet move2(Protocol::Opcode::PLAYER_POSITION);
            move2.writeByte(clientId);
            move2.writeByte(1); // Type
            move2.writeFloat(ox);
            move2.writeFloat(oy);
            move2.writeFloat(oz);
            move2.writeFloat(oyaw);
            move2.writeFloat(opitch);

            server->network->sendPacket(clientId, move2);

            return true;
        } else {
            Logger::logf(PREFIX_CC, "Can't teleport from console!\n");
        }
    }

    if (parts[0] == "/solid" && parts.size() == 1) {
        this->privateText = true;
        if (server->adminSystem->isAdmin(clientId) || console) {
            server->toggleSolid();
            if (server->getSolid()) {
                returnText = "Now placing unbreakable stone";
            } else {
                returnText = "Now placing normal stone";
            }
        } else {
            returnText = "&cYou cannot use this command.";
        }
        return true;
    }

    return false;
}


std::vector<std::string> CommandHandler::parseText(std::string& text) {
    std::stringstream ss(text);
    std::string token;
    std::vector<std::string> tokens;

    while (ss >> token) {
        tokens.push_back(token);
    }

    return tokens; // return splitted text (array)
}

std::string CommandHandler::getReturnText() {
    return this->returnText;
}

bool CommandHandler::getTextVisible() {
    return this->privateText; // true - only for client, who execute command ; false - for everyone
}